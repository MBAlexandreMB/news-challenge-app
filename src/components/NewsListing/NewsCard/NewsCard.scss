@import '../../../shared/styles/fonts.scss';
@import '../../../shared/styles/sizes.scss';
@import '../../../shared/styles/spacers.scss';

$card-height: 10rem;
$card-width: 39rem;

.news-card-container {
    display: flex;
    flex-flow: row nowrap;
    justify-content: start;
    box-sizing: border-box;
    max-width: $card-width;
    min-width: $card-width;
    height: $card-height;
    margin-bottom: $spacer-l;
    background-color: #fff;
    border-radius: $card-border-radius;
    cursor: pointer;

    &:not(:last-child) {
        margin-bottom: $margin-m;
    }
    
    .news-title {
        font-size: $font-size-l;
        font-weight: bold;
    }

    .news-image {
        max-width: 100px;
        min-width: 100px;
        width: 100px;
        height: $card-height;
        
        img {
            object-fit: cover;
            border-radius: $card-border-radius 0 0 $card-border-radius;
            width: 100%;
            height: 100%;
        }
    }

    .title-source-date-container {
        position: relative;
        width: 100%;
        height: $card-height;
        margin: $margin-xs $margin-m $margin-xs $margin-xs;
    }

    .news-source-date {
        position: absolute;
        bottom: $spacer-m;
        text-align: end;

        color: rgba(0, 0, 0, 0.3);
        font-style: italic;
    }
}

@media (max-width: #{$breakpoint-tablet}) {
    .news-card-container {
        width: 100%;
    }
}

@media (max-width: #{$breakpoint-small-devices}) {
    .news-card-container {
        min-width: 290px;

        .news-title {
            font-size: $font-size-m;
        }

        .news-source-date {
            font-size: $font-size-s;
        }
    }
}