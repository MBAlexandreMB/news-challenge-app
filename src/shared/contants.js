export const NO_IMAGE_URL = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRBJvWZNK1Lm7z-Iw0g4RJvw6l1sPwOUDOaZW_AecE3ZZdPyy4KpWejikv7VSD4cuUQpOY&usqp=CAU";