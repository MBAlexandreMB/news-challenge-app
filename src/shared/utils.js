import { NO_IMAGE_URL } from './contants';

export const onImageLoadingError = ({ currentTarget }) => {
    currentTarget.onerror = null;
    if (currentTarget.src !== NO_IMAGE_URL) {
        currentTarget.src= NO_IMAGE_URL;
    }
};