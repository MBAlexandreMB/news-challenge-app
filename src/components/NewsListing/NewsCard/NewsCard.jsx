import React from 'react';
import { DateTime } from 'luxon';

import { NO_IMAGE_URL, UNKNOWN_PUBLISHED_DATE, UNKNOWN_SOURCE } from '../../../shared/contants';
import { onImageLoadingError } from '../../../shared/utils';

import './NewsCard.scss';

const NewsCard = ({onSelectNews, news }) => {
    const { title, publishedAt } = news;
    const urlToImage = news.urlToImage || NO_IMAGE_URL;
    const sourceFromTitleIdx = (title && title.lastIndexOf("-")) || -1;

    let sourceFromTitle = UNKNOWN_SOURCE;
    let titleToUse = title;
    if (sourceFromTitleIdx !== -1) {
        sourceFromTitle = title.substring(sourceFromTitleIdx + 2);
        titleToUse = title.substring(0, sourceFromTitleIdx);
    }
    const publishedDate = publishedAt
        ? DateTime
            .fromISO(publishedAt)
            .toLocaleString(DateTime.DATETIME_SHORT)
        : UNKNOWN_PUBLISHED_DATE;

    return (
        <section className='news-card-container' onClick={() => onSelectNews(news)}>
            <figure className='news-image'>
                <img
                    src={urlToImage}
                    loading='lazy'
                    onError={onImageLoadingError}
                    alt='' // The API returns no description of the image
                />
            </figure>
            <div className="title-source-date-container">
                <h1 className='news-title'>{titleToUse}</h1>
                <p className="news-source-date">
                    <strong>{sourceFromTitle}</strong> | {publishedDate}
                </p>
            </div>
        </section>
    );
};
 
export default NewsCard;