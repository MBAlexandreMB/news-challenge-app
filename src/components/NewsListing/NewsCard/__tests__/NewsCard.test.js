import { shallow } from 'enzyme';
import { UNKNOWN_PUBLISHED_DATE, UNKNOWN_SOURCE } from '../../../../shared/contants';
import NewsCard from "../NewsCard";

const defaultProps = {
    onSelectNews: jest.fn(),
    news: [
        {
            title: 'testTitle - randomSource',
            publishedAt: new Date(2022, 0, 7, 2, 0).toISOString(),
            urlToImage: 'randomURL' }
        ],
};

describe('<NewsCard />', () => {
    it('should render correctly', () => {
        const wrapper = shallow(<NewsCard {...defaultProps} />);

        expect(wrapper).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should remove the source from the title', () => {
        const updatedNews = {
            ...defaultProps.news,
            title: 'test with source - source',
        };

        const wrapper = shallow(<NewsCard {...defaultProps} news={updatedNews} />);

        const title = wrapper.find('h1.news-title').text();
        const source = wrapper.find('[data-test="source"]').text();

        expect(title).toBe('test with source');
        expect(source).toBe('source');
    });

    it('should work with a title without a source', () => {
        const updatedNews = {
            ...defaultProps.news,
            title: 'test without source',
        };

        const wrapper = shallow(<NewsCard {...defaultProps} news={updatedNews} />);

        const title = wrapper.find('h1.news-title').text();
        const source = wrapper.find('[data-test="source"]').text();

        expect(title).toBe('test without source');
        expect(source).toBe(UNKNOWN_SOURCE);
    });

    it('should set the published date', () => {
        const updatedNews = {
            ...defaultProps.news,
            publishedAt: new Date(2022, 0, 9, 2, 0).toISOString(),
        };

        const wrapper = shallow(<NewsCard {...defaultProps} news={updatedNews} />);

        const publishedAt = wrapper.find('[data-test="publishedAt"]').text();

        expect(publishedAt).toBe('09/01/2022, 02:00');
    });

    it('should work without a published date', () => {
        const updatedNews = {
            ...defaultProps.news,
            publishedAt: undefined,
        };

        const wrapper = shallow(<NewsCard {...defaultProps} news={updatedNews} />);

        const publishedAt = wrapper.find('[data-test="publishedAt"]').text();

        expect(publishedAt).toBe(UNKNOWN_PUBLISHED_DATE);
    });
    
    it('call onSelectNews when clicked', () => {
        const wrapper = shallow(<NewsCard {...defaultProps} />);

        wrapper.find('[data-test="card-container"]').simulate('click');

        expect(defaultProps.onSelectNews).toHaveBeenCalled();
    });
});