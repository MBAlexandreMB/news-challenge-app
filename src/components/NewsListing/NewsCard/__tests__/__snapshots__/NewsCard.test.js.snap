// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsCard /> should render correctly 1`] = `
<section
  className="news-card-container"
  data-test="card-container"
  onClick={[Function]}
>
  <figure
    className="news-image"
  >
    <img
      alt=""
      loading="lazy"
      onError={[Function]}
      src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRBJvWZNK1Lm7z-Iw0g4RJvw6l1sPwOUDOaZW_AecE3ZZdPyy4KpWejikv7VSD4cuUQpOY&usqp=CAU"
    />
  </figure>
  <div
    className="title-source-date-container"
  >
    <h1
      className="news-title"
    />
    <p
      className="news-source-date"
    >
      <strong
        data-test="source"
      >
        Unknown
      </strong>
       | 
      <span
        data-test="publishedAt"
      >
        Unknown publication date
      </span>
    </p>
  </div>
</section>
`;
